File: service/resource-encapsulation/src/ResourceClient.cpp
//******************************************************************
//
// Copyright 2015 Samsung Electronics All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

#include "ResourceClient.h"
#include "ResourceBroker.h"
#include "ResourceCacheManager.h"

#define CLIENT_W_TAG  PCF("ResourceClient")

using namespace OIC::Service;

namespace
{
    ResourceState getResourceStateFromBrokerState(BROKER_STATE state)
    {

        OC_LOG(DEBUG, CLIENT_W_TAG, "getResourceStateFromBrokerState entry");

        if (state == BROKER_STATE::ALIVE)
        {
            return ResourceState::ALIVE;
        }
        else if (state == BROKER_STATE::REQUESTED)
        {
            return ResourceState::REQUESTED;
        }
        else if (state == BROKER_STATE::LOST_SIGNAL)
        {
            return ResourceState::LOST_SIGNAL;
        }
        else if (state == BROKER_STATE::DESTROYED)
        {
            return ResourceState::DESTROYED;
        }
        else if (state == BROKER_STATE::NONE)
        {
            return ResourceState::NOT_MONITORING;
        }
        OC_LOG(ERROR, CLIENT_W_TAG, "getResourceStateFromBrokerState ERROR");

        //Default return value
        return ResourceState::NOT_MONITORING;
    }

    CacheState getCacheState(CACHE_STATE state)
    {

        OC_LOG(DEBUG, CLIENT_W_TAG, "getCacheState (from CACHE_STATE) entry");

        if (state == CACHE_STATE::READY)
        {
            return CacheState::READY;
        }
        else if (state == CACHE_STATE::READY_YET)
        {
            return CacheState::READY_YET;
        }
        else if (state == CACHE_STATE::LOST_SIGNAL)
        {
            return CacheState::LOST_SIGNAL;
        }
        else if (state == CACHE_STATE::DESTROYED)
        {
            return CacheState::DESTROYED;
        }
        else if (state == CACHE_STATE::UPDATING)
        {
            return CacheState::UPDATING;
        }
        else if (state == CACHE_STATE::NONE)
        {
            return CacheState::NONE;
        }
        OC_LOG(ERROR, CLIENT_W_TAG, "getCacheState (from CACHE_STATE) ERROR");

        //Default return value
        return CacheState::NONE;
    }

    OCStackResult hostingCallback(BROKER_STATE state,
                                  RemoteResourceObject::ResourceStateChangedCallback onResourceStateChanged)
    {
        OC_LOG(DEBUG, CLIENT_W_TAG, "hostingCallback entry");

        ResourceState resourceState = getResourceStateFromBrokerState(state);
        onResourceStateChanged(resourceState); //passing ResourceState to application

        OC_LOG(DEBUG, CLIENT_W_TAG, "hostingCallback exit");
        return OC_STACK_OK;
    }

    OCStackResult cachingCallback(std::shared_ptr<PrimitiveResource> resource,
                                  const ResourceAttributes &data,
                                  RemoteResourceObject::CacheUpdatedCallback onCacheUpdated)
    {
        OC_LOG(DEBUG, CLIENT_W_TAG, "cachingCallback entry");

        onCacheUpdated(data); //passing ResourceAttributes to application

        OC_LOG(DEBUG, CLIENT_W_TAG, "cachingCallback exit");
        return OC_STACK_OK;
    }

    void setCallback(const HeaderOptions &header, const ResponseStatement &response, int n,
                     RemoteResourceObject::RemoteAttributesSetCallback onRemoteAttributesSet)
    {
        OC_LOG(DEBUG, CLIENT_W_TAG, "setCallback entry");

        const ResourceAttributes &attributes = response.getAttributes();
        onRemoteAttributesSet(attributes); //passing ResourceAttributes to application

        OC_LOG(DEBUG, CLIENT_W_TAG, "setCallback exit");
    }

    void getCallback(const HeaderOptions &headerOption, const ResponseStatement &response, int n,
                     RemoteResourceObject::RemoteAttributesReceivedCallback onRemoteAttributesReceived)
    {
        OC_LOG(DEBUG, CLIENT_W_TAG, "getCallback entry");

        const ResourceAttributes &attributes = response.getAttributes();
        onRemoteAttributesReceived(attributes); //passing ResourceAttributes to application

        OC_LOG(DEBUG, CLIENT_W_TAG, "getCallback exit");
    }

    void findCallback(std::shared_ptr<PrimitiveResource> primitiveResource,
                      DiscoveryManager::OnResourceDiscoveredCallback OnResourceDiscovered )
    {
        OC_LOG(DEBUG, CLIENT_W_TAG, "findCallback entry");

        if (nullptr == primitiveResource)
        {
            OC_LOG(ERROR, CLIENT_W_TAG, "findCallback::primitiveResource NULL Parameter");
            return ;
        }

        std::shared_ptr< RemoteResourceObject>  primitiveClientResource =
            std::shared_ptr< RemoteResourceObject>(new RemoteResourceObject(primitiveResource));

        OnResourceDiscovered(primitiveClientResource); //passing PrimitiveClientResource to application

        OC_LOG(DEBUG, CLIENT_W_TAG, "findcb exit");
    }

}

//*******************************Primitive Client Resource*************************************

namespace OIC
{
    namespace Service
    {

        RemoteResourceObject:: RemoteResourceObject(std::shared_ptr<PrimitiveResource>  pResource) :
            m_monitoringFlag(false), m_cachingFlag(false),  m_observableFlag(pResource->isObservable()),
            m_primitiveResource(pResource), m_cacheId(0), m_brokerId(0) {}

        bool RemoteResourceObject::isMonitoring() const
        {
            return m_monitoringFlag;
        }

        bool RemoteResourceObject::isCaching() const
        {
            return m_cachingFlag;
        }

        void RemoteResourceObject::startMonitoring(ResourceStateChangedCallback cb)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startMonitoring entry");
            if (true == m_monitoringFlag)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startMonitoring : Already started");
            }
            else
            {
                try
                {
                    BrokerID brokerId =  ResourceBroker::getInstance()->hostResource(m_primitiveResource,
                                         std::bind(hostingCallback, std::placeholders::_1,
                                                   cb));
                    m_monitoringFlag = true;
                    m_brokerId = brokerId;
                }
                catch (std::exception &exception)
                {
                    throw InvalidParameterException {exception.what()};
                }
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startMonitoring exit");
        }

        void RemoteResourceObject::stopMonitoring()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::stopMonitoring entry");
            if (true == m_monitoringFlag)
            {
                try
                {
                    ResourceBroker::getInstance()->cancelHostResource(m_brokerId);
                    m_monitoringFlag = false;
                }
                catch (std::exception &exception)
                {
                    OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::stopMonitoring InvalidParameterException");
                }
            }
            else
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject:: stopMonitoring : already terminated");
            }

            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::stopMonitoring exit");
        }

        ResourceState RemoteResourceObject::getState() const
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, " RemoteResourceObject::getState entry");
            try
            {
                BROKER_STATE brokerState = ResourceBroker::getInstance()->getResourceState(m_primitiveResource);
                OC_LOG(DEBUG, CLIENT_W_TAG, " RemoteResourceObject::getState exit");
                return getResourceStateFromBrokerState(brokerState);
            }
            catch (std::exception &exception)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, " RemoteResourceObject::getState InvalidParameterException");
                throw BadRequestException { "[getState] Get Resource Source State from Broker Error " };
            }
        }

        void RemoteResourceObject::startCaching()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching entry");
            if (true == m_cachingFlag)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching : already Started");
            }
            else
            {
                try
                {
                  CacheID cacheId = ResourceCacheManager::getInstance()->requestResourceCache(m_primitiveResource,
                                      NULL, REPORT_FREQUENCY::NONE,  0);

                    m_cacheId = cacheId;
                    m_cachingFlag = true;
                    OC_LOG_V(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching CACHE ID %d", cacheId);
                }
                catch (std::exception &exception)
                {
                    OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching InvalidParameterException");
                }
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching exit");
        }

        void RemoteResourceObject::startCaching(CacheUpdatedCallback cb)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching entry");
            if (!cb)
            {
                throw InvalidParameterException {"startCaching : Callback is NULL" };
            }
            else
            {
                if (true == m_cachingFlag)
                {
                    OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching : already Started");
                }
                else
                {
                    try
                    {
                        CacheID cacheId = ResourceCacheManager::getInstance()->requestResourceCache(m_primitiveResource,
                                          std::bind(cachingCallback, std::placeholders::_1, std::placeholders::_2, cb),
                                          REPORT_FREQUENCY::UPTODATE,  0);

                        m_cacheId = cacheId;
                        m_cachingFlag = true;
                        OC_LOG_V(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching CACHE ID %d", cacheId);
                    }
                    catch (std::exception &exception)
                    {
                        throw InvalidParameterException {"startCaching : error" };
                    }
                }
            }
        }

        void RemoteResourceObject::stopCaching()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::stopCaching entry");

            if (true == m_cachingFlag)
            {
                try
                {
                    ResourceCacheManager::getInstance()->cancelResourceCache(m_cacheId);
                    m_cachingFlag = false;
                }
                catch (std::exception &exception)
                {
                    OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::stopCaching InvalidParameterException");
                }
            }
            else
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject:: Caching already terminated");
            }

            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::stopCaching exit");
        }

        CacheState  RemoteResourceObject::getResourceCacheState()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::getResourceCacheState entry");
            try
            {
                CACHE_STATE cacheState = ResourceCacheManager::getInstance()->getResourceCacheState(
                                             m_primitiveResource);
                OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::getResourceCacheState exit");
                return getCacheState(cacheState);
            }
            catch (std::exception &exception)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG,
                       "RemoteResourceObject::getResourceCacheState InvalidParameterException");
                throw BadRequestException { "[getResourceCacheState] Caching not started" };
            }
        }

        void RemoteResourceObject::refreshCache()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::refreshCache entry");
            try
            {

                ResourceCacheManager::getInstance()->updateResourceCache(
                    m_primitiveResource);
            }
            catch (std::exception &exception)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::refreshCache InvalidParameterException");
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::refreshCache exit");
        }

        ResourceAttributes RemoteResourceObject:: getCachedAttributes() const
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject :: getCachedAttributes ");
            try
            {
                return ResourceCacheManager::getInstance()->getCachedData(m_primitiveResource);
            }
            catch (std::exception &exception)
            {
                throw BadRequestException { "[getCachedAttributes]  Caching not started" };
            }
        }

        ResourceAttributes::Value  RemoteResourceObject:: getCachedAttribute( const std::string &key)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject :: getCachedAttribute entry");
            try
            {
                ResourceAttributes  Cachedattributes = ResourceCacheManager::getInstance()->getCachedData(
                        m_primitiveResource);
                return Cachedattributes[key];
            }
            catch (std::exception &exception)
            {
                throw BadRequestException { "[getCachedAttribute]  Caching not started" };
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject :: getCachedAttribute exit");
        }

        std::string RemoteResourceObject::getUri() const
        {
            return m_primitiveResource->getUri();
        }

        std::string RemoteResourceObject::getAddress() const
        {
            return m_primitiveResource->getHost();
        }

        std::vector < std::string > RemoteResourceObject::getTypes() const
        {
            return m_primitiveResource->getTypes();
        }

        std::vector < std::string > RemoteResourceObject::getInterfaces() const
        {
            return m_primitiveResource->getInterfaces();
        }

        void RemoteResourceObject::getRemoteAttributes(RemoteAttributesReceivedCallback cb)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::getRemoteAttributes entry");
            if (!cb)
            {
                throw InvalidParameterException {"getRemoteAttributes : Callback is NULL" };
            }
            else
            {
                m_primitiveResource->requestGet(std::bind(getCallback, std::placeholders::_1,
                                                std::placeholders::_2, std::placeholders::_3, cb));
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::getRemoteAttributes exit");
        }

        void RemoteResourceObject::setRemoteAttributes(ResourceAttributes &attribute,
                RemoteAttributesSetCallback cb)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::setRemoteAttributes entry");
            if (!cb)
            {
                throw InvalidParameterException {"setRemoteAttributes : Callback is NULL" };
            }
            else
            {
                m_primitiveResource->requestSet(attribute, std::bind(setCallback, std::placeholders::_1,
                                                std::placeholders::_2, std::placeholders::_3, cb));
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::setRemoteAttributes exit");
        }

        bool RemoteResourceObject::isObservable() const
        {
            return m_observableFlag;
        }

//*******************************Discovery Manager****************************************

        DiscoveryManager *DiscoveryManager:: getInstance()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "DiscoveryManager:: getInstance entry");
            static DiscoveryManager *s_instance;
            static std::mutex s_mutex;
            if (!s_instance)
            {
                std::lock_guard<std::mutex> lock(s_mutex);
                if (!s_instance)
                {
                    s_instance = new DiscoveryManager();
                }
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "DiscoveryManager:: getInstance exit");
            return s_instance;
        }

        void DiscoveryManager::discoverResource(const RCSAddress &address, const std::string &resourceURI,
                                                OnResourceDiscoveredCallback cb)
        {

            OC_LOG(DEBUG, CLIENT_W_TAG, "DiscoveryManager::discoverResource entry");

            if ( resourceURI.empty() )
            {
                OC_LOG(ERROR, CLIENT_W_TAG, "discoverResource NULL resourceURI");
                throw InvalidParameterException { "discoverResource NULL resourceURI'" };
            }
            else if ( !cb )
            {
                OC_LOG(ERROR, CLIENT_W_TAG, "discoverResource NULL Callback");
                throw InvalidParameterException { "discoverResource NULL Callback'" };
            }
            OIC::Service::discoverResource(address, resourceURI, std::bind(findCallback,
                                           std::placeholders::_1,
                                           cb));

            OC_LOG(DEBUG, CLIENT_W_TAG, "DiscoveryManager::discoverResource exit");
        }
    }
}

File: service/resource-encapsulation/examples/android/SampleResourceClientApp/src/com/re/sampleclient/ResourceClient.java
/******************************************************************
 *
 * Copyright 2015 Samsung Electronics All Rights Reserved.
 *
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************/
package com.re.sampleclient;

import java.util.Vector;

import org.iotivity.ResourceEncapsulation.client.RCSAddress;
import org.iotivity.ResourceEncapsulation.client.RCSDiscoveryManager;
import org.iotivity.ResourceEncapsulation.client.RCSDiscoveryManager.IDiscoverResourceListener;
import org.iotivity.ResourceEncapsulation.client.RCSRemoteResourceObject.IStateChangedCallbackListener;
import org.iotivity.ResourceEncapsulation.client.RCSRemoteResourceObject.ICacheUpdateListener;
import org.iotivity.ResourceEncapsulation.client.RCSRemoteResourceObject.IRemoteAttributeListener;
import org.iotivity.ResourceEncapsulation.client.RCSRemoteResourceObject.CacheState;
import org.iotivity.ResourceEncapsulation.client.RCSRemoteResourceObject.ResourceState;
import org.iotivity.ResourceEncapsulation.client.RCSRemoteResourceObject;
import org.iotivity.ResourceEncapsulation.common.RCSException;
import org.iotivity.ResourceEncapsulation.common.RCSResourceAttributes;

import android.os.Message;
import android.util.Log;

/**
 * For calling the Resource Client APIs as per user selection on UI and for
 * updating the UI
 *
 * It contains all the Resource Client APIs.
 */
public class ResourceClient {

    private static final String            LOG_TAG                  = "[RESample] Client";
    private static Message                 msg;
    public String                          logMessage;

    private final RCSDiscoveryManager      discoveyManagerObj;
    public RCSRemoteResourceObject         remoteResourceObj;

    private static ResourceClientActivity  resourcClientActivityObj = null;
    private final DiscoverResourceListener discoverResourceListener;
    private final StateChangedListener     stateChangedListener;
    private final CacheUpdateListener      cacheUpdateListener;
    private final RemoteAttributeListener  remoteAttributeListener;
    static boolean                         monitoringState          = false;
    static boolean                         cachingState             = false;

    // constructor
    public ResourceClient() {
        discoveyManagerObj = RCSDiscoveryManager.getInstance();
        discoverResourceListener = new DiscoverResourceListener();
        stateChangedListener = new StateChangedListener();
        cacheUpdateListener = new CacheUpdateListener();
        remoteAttributeListener = new RemoteAttributeListener();

        discoveyManagerObj
                .setDiscoverResourceListener(discoverResourceListener);

        resourcClientActivityObj = ResourceClientActivity
                .getResourceClientActivityObj();
    }

    /**
     * Listener for receiving Resource discovered in network.
     */
    private class DiscoverResourceListener implements IDiscoverResourceListener {

        @Override
        public void onResourceDiscovered(RCSRemoteResourceObject foundResource) {
            Log.i(LOG_TAG, "onResourceDiscovered invoked");

            String resourceURI = foundResource.getUri();
            String hostAddress = foundResource.getAddress();
            boolean isObservableflag = foundResource.isObservable();

            Vector<String> resourceTypes = foundResource.getTypes();
            Vector<String> resourceInterfaces = foundResource.getInterFaces();
            remoteResourceObj = foundResource;

            msg = Message.obtain();
            msg.what = 0;
            resourcClientActivityObj.getHandler().sendMessage(msg);

            logMessage = "URI : " + resourceURI + "\n";
            logMessage = logMessage + "Host : " + hostAddress + "\n";
            for (int i = 0; i < resourceTypes.size(); i++) {
                logMessage = logMessage + "resourceType : "
                        + resourceTypes.get(i) + "\n";
            }

            for (int i = 0; i < resourceInterfaces.size(); i++) {
                logMessage = logMessage + "resourceInterfaces : "
                        + resourceInterfaces.get(i) + "\n";
            }
            logMessage = logMessage + "isObservable : " + isObservableflag
                    + "\n";

            ResourceClientActivity.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 1;
            resourcClientActivityObj.getHandler().sendMessage(msg);
        }
    };

    /**
     * Listener for receiving Changed resource state as a result of
     * startMonitoring API
     */
    private class StateChangedListener implements IStateChangedCallbackListener {

        @Override
        public void onStateChanged(ResourceState resourceState) {
            Log.i(LOG_TAG, "onStateChangedCallback invoked");

            switch (resourceState) {
                case NONE:
                    Log.i(LOG_TAG, "Current Resource State : NOT_MONITORING");
                    logMessage = "Current Resource State : NOT_MONITORING\n";
                    break;

                case ALIVE:
                    Log.i(LOG_TAG, "Current Resource State : ALIVE");
                    logMessage = "Current Resource State : ALIVE\n";
                    break;

                case REQUESTED:
                    Log.i(LOG_TAG, "Current Resource State : REQUESTED");
                    logMessage = "Current Resource State : REQUESTED\n";
                    break;

                case LOST_SIGNAL:
                    Log.i(LOG_TAG, "Current Resource State : LOST_SIGNAL");
                    logMessage = "Current Resource State : LOST_SIGNAL\n";
                    logMessage = logMessage
                            + "Start the server and press discover resource button"
                            + "\n";
                    break;

                case DESTROYED:
                    Log.i(LOG_TAG, "Current Resource State : DESTROYED");
                    logMessage = "Current Resource State : DESTROYED\n";
                    break;
            }

            ResourceClientActivity.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 1;
            resourcClientActivityObj.getHandler().sendMessage(msg);
        }

    }

    /**
     * Listener for receiving updated resource attributes as a result of
     * startCaching() API with callback
     */
    private class CacheUpdateListener implements ICacheUpdateListener {
        @Override
        public void onCacheUpdatedCallback(RCSResourceAttributes attributes) {
            Log.i(LOG_TAG, "onCacheUpdatedCallback invoked");

            printAttributes(attributes);
        }
    };

    /**
     * Listener for receiving the resource attributes as a result of
     * getRemoteAttributes() and setRemoteAttributes() API
     */
    private class RemoteAttributeListener implements IRemoteAttributeListener {
        @Override
        public void onRemoteAttributesGetCallback(
                RCSResourceAttributes attributes) {

            Log.i(LOG_TAG, "onRemoteAttributesGetCallback invoked");
            printAttributes(attributes);
        }
    };

    // Discover Resource
    public void disocverResources() {
        Log.d(LOG_TAG, "discover resources entry");
        RCSAddress address = RCSAddress.multicast();
        try {
            discoveyManagerObj.discoverResource(address);
        } catch (RCSException e) {
            e.printStackTrace();
        }
        Log.d(LOG_TAG, "discover resources exit");
    }

    // Start Monitoring
    public void startMonitoring() {
        Log.d(LOG_TAG, "startMonitoring entry");

        if (null == remoteResourceObj) {
            Log.e(LOG_TAG, "Remote Resource object NULL");
            return;
        }

        if (true == monitoringState) {
            logMessage = "Monitoring already started\n";
            Log.e(LOG_TAG, "Monitoring already started");
            ResourceClientActivity.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 1;
            resourcClientActivityObj.getHandler().sendMessage(msg);
            return;
        }

        remoteResourceObj.setStateChangedListener(stateChangedListener);

        if (null != stateChangedListener) {
            try {
                remoteResourceObj.startMonitoring();
                monitoringState = true;
            } catch (RCSException e) {
                e.printStackTrace();
                return;
            }
        }
        Log.d(LOG_TAG, "startMonitoring exit");
    }

    // Stop Monitoring
    public void stopMonitoring() {
        Log.d(LOG_TAG, "stopMonitoring entry");

        if (monitoringState) {
            if (null == remoteResourceObj) {
                Log.e(LOG_TAG, "Remote Resource object NULL");
                return;
            }

            remoteResourceObj.stopMonitoring();
            monitoringState = false;
            logMessage = "Stopped Resource Monitoring\n";
        } else {
            Log.i(LOG_TAG, "Monitoring not started");
            logMessage = "Monitoring not started\n";
        }

        Log.d(LOG_TAG, "stopMonitoring exit");

        ResourceClientActivity.setMessageLog(logMessage);
        msg = Message.obtain();
        msg.what = 1;
        resourcClientActivityObj.getHandler().sendMessage(msg);
    }

    // Get Attributes
    public void getRemoteAttributes() {
        Log.d(LOG_TAG, "getRemoteAttributes entry");

        if (null == remoteResourceObj) {
            Log.e(LOG_TAG, "Remote Resource object NULL");
            return;
        }

        try {
            remoteResourceObj
                    .setRemoteAttributesListener(remoteAttributeListener);

            if (null != remoteAttributeListener) {
                remoteResourceObj.getRemoteAttributes();
            }
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }

        Log.d(LOG_TAG, "getRemoteAttributes exit");
    }

    // Set Attributes
    public void setRemoteAttributes(int value) {
        Log.d(LOG_TAG, "setRemoteAttributes entry");

        if (null == remoteResourceObj) {
            Log.e(LOG_TAG, "Remote Resource object NULL");
            return;
        }

        try {
            remoteResourceObj
                    .setRemoteAttributesListener(remoteAttributeListener);

            if (null != remoteAttributeListener) {
                RCSResourceAttributes attr = new RCSResourceAttributes();
                attr.setValueInt("Temperature", value);
                remoteResourceObj.setRemoteAttributes(attr);
            }
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }

        Log.d(LOG_TAG, "setRemoteAttributes exit");
    }

    // Start Caching
    public void startCaching(int cachingType) {
        Log.d(LOG_TAG, "startCaching entry");

        if (null == remoteResourceObj) {
            Log.e(LOG_TAG, "Remote Resource object NULL");
            return;
        }

        if (true == cachingState) {
            Log.d(LOG_TAG, "Caching already started");
            logMessage = "Caching already started\n";
            ResourceClientActivity.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 1;
            resourcClientActivityObj.getHandler().sendMessage(msg);
            return;
        }

        if (0 == cachingType) {
            Log.d(LOG_TAG, "startCaching without callback");
            remoteResourceObj.setCacheUpdateListener(null);
            logMessage = "Caching started without callback\n";
            ResourceClientActivity.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 1;
            resourcClientActivityObj.getHandler().sendMessage(msg);
        } else {
            Log.d(LOG_TAG, "startCaching with callback");
            remoteResourceObj.setCacheUpdateListener(cacheUpdateListener);
            logMessage = "Caching started\n";
            ResourceClientActivity.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 1;
            resourcClientActivityObj.getHandler().sendMessage(msg);
        }

        try {
            remoteResourceObj.startCaching();
            cachingState = true;
        } catch (RCSException e) {
            e.printStackTrace();
            return;
        }

        Log.d(LOG_TAG, "startCaching exit");
    }

    // Get Cache State
    public void getCacheState() {
        Log.d(LOG_TAG, "getCacheState entry");

        if (null == remoteResourceObj) {
            Log.e(LOG_TAG, "Remote Resource object NULL");
            return;
        }

        CacheState cState = remoteResourceObj.getCacheState();
        Log.i(LOG_TAG, "Cache State : " + cState.name());

        logMessage = "Cache State : " + cState.toString() + "\n";
        ResourceClientActivity.setMessageLog(logMessage);
        msg = Message.obtain();
        msg.what = 1;
        resourcClientActivityObj.getHandler().sendMessage(msg);

        Log.d(LOG_TAG, "getCacheState exit");
    }

    // Get Cached Attributes
    public void getCachedAttributes() {
        Log.d(LOG_TAG, "getCachedAttributes entry");

        if (null == remoteResourceObj) {
            Log.e(LOG_TAG, "Remote Resource object NULL");
            return;
        }
        if (true == cachingState) {
            RCSResourceAttributes attributes = null;
            try {
                attributes = remoteResourceObj.getCachedAttributes();
            } catch (RCSException e) {
                e.printStackTrace();
                return;
            }
            printAttributes(attributes);
        } else {
            logMessage = "Caching not started" + "\n";
            ResourceClientActivity.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 1;
            resourcClientActivityObj.getHandler().sendMessage(msg);
        }
        Log.d(LOG_TAG, "getCachedAttributes exit");
    }

    // Get Cached Attribute
    public void getCachedAttribute() {
        Log.d(LOG_TAG, "getCachedAttribute entry");

        if (null == remoteResourceObj) {
            Log.e(LOG_TAG, "Remote Resource object NULL");
            return;
        }

        if (true == cachingState) {
            String key = "Temperature";
            String attribute;
            try {
                attribute = remoteResourceObj.getCachedAttribute(key);
            } catch (RCSException e) {
                e.printStackTrace();
                return;
            }
            logMessage = "Key : " + key + "\n";
            logMessage = logMessage + "Value : " + attribute + "\n";
            ResourceClientActivity.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 1;
            resourcClientActivityObj.getHandler().sendMessage(msg);
        } else {
            logMessage = "Caching not started" + "\n";
            ResourceClientActivity.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 1;
            resourcClientActivityObj.getHandler().sendMessage(msg);
        }
        Log.d(LOG_TAG, "getCachedAttribute exit");
    }

    // Stop Caching
    public void stopCaching() {
        Log.d(LOG_TAG, "stopCaching entry");

        if (null == remoteResourceObj) {
            Log.e(LOG_TAG, "Remote Resource object NULL");
            return;
        }

        if (false == cachingState) {
            Log.e(LOG_TAG, "Caching not started");
            logMessage = "Caching not started\n";
        } else {
            remoteResourceObj.stopCaching();
            cachingState = false;
            logMessage = "Stopped Caching\n";
        }

        ResourceClientActivity.setMessageLog(logMessage);
        msg = Message.obtain();
        msg.what = 1;
        resourcClientActivityObj.getHandler().sendMessage(msg);
        Log.d(LOG_TAG, "stopCaching exit");
    }

    // For Printing the Attributes on the UI
    private void printAttributes(RCSResourceAttributes attributes) {
        Log.d(LOG_TAG, "Printing Attributes");

        try {
            int temperature = attributes.getValueInt("Temperature");
            Log.i(LOG_TAG, "Temperature : " + temperature);
            logMessage = "Attributes : \n\nTemperature : " + temperature + "\n";
        } catch (Exception e) {
            Log.i(LOG_TAG, "Temperature attribute not found");
            logMessage = "Attributes Temperature not found \n";
            e.printStackTrace();
            return;
        }
        ResourceClientActivity.setMessageLog(logMessage);
        msg = Message.obtain();
        msg.what = 1;
        resourcClientActivityObj.getHandler().sendMessage(msg);
    }
}